### DISCUSSION TITLE
Paper or Documentation on neighbors.pyx

### DISCUSSION THREAD

**Discussion by rydeveraumn** [materialsproject/pymatgen#3500] [General]
Hello all,

Being fairly new to materials science and `pymatgen` I have been working through the details of the `MEGNet` paper and model. It seems a key detail is an execution of https://github.com/materialsproject/pymatgen/blob/master/pymatgen/optimization/neighbors.pyx. I would like to get a better, lower level understanding and am having trouble understanding how the lattice comes into play and how the offsets are being created. Is there any literature that describes this process?

**Comment by janosh** [comment]
I think @lbluque and @AegisIK were the ones to most recently take a deep dive on that part of the code and may be able to give you some pointers.

**Reply by shyuep** [reply]
There is no paper on this. We did it on the way to making MEGNEt faster. It is an algorithmic improvement, not a scientific one. Too trivial for us to publish.

**Reply by rydeveraumn** [reply]
Hello @janosh and @shyuep! Thank you so much for the fast response. I think my wording may have been too specific so let me ask some clarifying questions.

1. Is this logic that is specific to the `MEGNet` `M3GNet` models? If so then it sounds like I might have to dig through the source code to understand exactly what is happening because it seems a little more than just finding the neighbors within some radius is taking place. i.e, what are the offset vectors, pbc? and why are there more outputs than cartesian coordinates for mp-1215 for example?

2. If this is common logic within the field of materials science, solid state physics, etc., then are there any recommendations for resources to fill in my knowledge gaps? Sounds like I should also read through references within the papers.

Since the features of the graph (mainly) come from this calculation it feels like an important part of the modeling step to fully comprehend because data representation to the model plays a key role in what it can learn. Let me know if you may have different conclusions.

Warm regards,

Ryan

**Reply by rydeveraumn** [reply]
BTW are these functions equivalent with the difference being one is cythonized? https://github.com/materialsproject/pymatgen/blob/v2023.10.4/pymatgen/core/lattice.py#L1730-L1855 if so I can dissect this function 🙂
